regressions fixes

emonget opened the pull request.

emonget committed:
fix: regressions + misc refactoring

emonget committed:
fix(boards): reduce patch footprint

emonget committed:
fix: add safe access to Biome singleton to prevent uninitialized use

emonget committed:
fix: misc regressions

emonget committed:
feat: add fake empty chunks gen option

emonget committed:
fix: board items block encoding

emonget committed:
refactor: move demo configs to test folder

emonget committed:
misc: adjust biome repartition

emonget committed:
feat: add non regression tests

emonget committed:
fix: types, formatting, unwanted committed files

emonget committed:
fix: handle task cancelling properly

Sceat commented on src/node/NodeWorkerProxy.ts:
```diff
@@ -1,15 +1,23 @@
+  override init(worldEnv: WorldEnv): Promise<any> {
     // node env
-    const workerUrl = new URL('./world_compute_node_worker.js', import.meta.url)
+    const workerUrl = new URL('./world_compute_node_worker', import.meta.url)
```
This is a delicate change, has it been tested with both a packaging (`npm pack`) in both nodejs and the dApp ?

emonget replied:
probably unwanted line coming from bad copy/pasting, as I had to dig changes from old branch

Sceat commented on src/processing/BoardProvider.ts:
```diff
@@ -257,9 +259,9 @@ export class BoardProvider {
     boardThickness?: number,
   ) {
-    boardRadius = boardRadius || worldEnv.rawSettings.boards.boardRadius
+    boardRadius = boardRadius || worldRootEnv.rawSettings.boards.boardRadius
```
We do not want default settings for boards either, is the worldRootEnv something that must be defined ? The point is creating a board requires specific settings and they should always be provided specifically

emonget replied:
changed to make these params mandatory, meaning users will need to provide them every time a new board is created. Initially, they were optional because I assumed they wouldn't change frequently. Additionally, their default values weren't hardcoded, but instead came from globally shared settings, which was intended to simplify the world setup by centralizing all the default settings in one place

Sceat commented on src/processing/BoardProvider.ts:
```diff
@@ -276,12 +278,15 @@ export class BoardProvider {
     return getPatchId(
       asVect2(this.boardParams.center),
-      worldEnv.getPatchDimensions(),
+      worldRootEnv.getPatchDimensions(),
```
Can this also be given in the constructor of the board ? To avoid using configuration which we did not explicitly define (and thus could be false)

emonget replied:
this has to be consistant everywhere in the world, you cannot provide different values for each component which would lead to unexpected issues, hence the need to have a common configuration shared across all world modules.
if we decide to have multiple world instance later this will still be shared between all world modules inside that instance.

Sceat commented on src/processing/ChunksPolling.ts:
```diff
@@ -2,22 +2,18 @@ import { Vector2 } from 'three'
-const chunksRange = worldEnv.rawSettings.chunks.range
-const { patchViewRanges } = worldEnv.rawSettings
+const chunksRange = worldRootEnv.rawSettings.chunks.range
+const { patchViewRanges } = worldRootEnv.rawSettings
```
Root scope usage of the default config will lead to discrepancies

emonget replied:
that's right, forgot to remove it from root scope here

Sceat commented on src/procgen/Biome.ts:
```diff
@@ -196,14 +211,41 @@ export class Biome {
+    if (!isEmptyConf) {
+      this.parseBiomesConfig(Biome.externalRawConf)
+    } else {
+      console.warn(
+        `missing biome configuration at creation, must provide conf later`,
+      )
```
A warning is not useful and only for human consumption, we want to strictly fail if there is no configuration

Sceat commented on src/procgen/Biome.ts:
```diff
@@ -196,14 +211,41 @@ export class Biome {
 
   static get instance() {
     return (Biome.singleton ||= new Biome())
   }
 
-  ensureInitialized() {
-    if (!this.#initialized) throw new Error('Biome not initialized')
+  // safe access from outside of the world to prevent uninitialized use
+  static get safeInstance() {
+    return Biome.singleton?.ready ? Biome.singleton : null
+  }
```
in which scenario would we want to use an unsafe instance ? (by calling `instance`) why not simply only using safe instance and having a simpler usage, this is increased complexity

emonget replied:
to be able to call this: `parseBiomesConfig` for instance?
On demand init is inherent to singleton pattern, you need to get the instance first and call the init afterwards.

Sceat commented on src/procgen/Biome.ts:
```diff
@@ -196,14 +211,41 @@ export class Biome {
+    return Biome.singleton?.ready ? Biome.singleton : null
+  }
+
+  get ready() {
+    return Object.keys(this.mappings).length > 0
   }
```
So a biome can be considered "ready" which is an additional step and will still partially work when not ready, instead of increasing the complexity we could go straight to the point and have it very simple and straightforward, `new Biome(config)` and nothing else, no need for additional checks/concepts

emonget replied:
Such design change are out of scope of this PR which mainly focus on fixing regression, 
When functional requirements changes arise, we can decide if using singleton pattern for Biome is still relevant and if we want to support multiple Biome instance at the same time.

Sceat reviewed (changes requested):
This pr seem to contain breaking changes, could you add a description to it to define what should be changed in other services using the world?

emonget committed:
feat: making all board parameters mandatory

emonget committed:
misc: requested changes

emonget committed:
fix: formatting

emonget committed:
fix: missed build errors

emonget commented:
Among changes that can affect external world use:
1. world env init: 
- from main thread: worldRootEnv.fromStub(world_env)
- from workerpool: initPoolEnv(world_env)
2. added `task.processingParams.fakeEmpty` flag 
3. added `safeInstance` static getter for direct access from outside the world

For better overview of changes,  you can also check non-regression tests in `tests`folder, which contain most common world usage examples (will be completed over time)

emonget commented:
closed in favor of #43
